### PR TITLE
DFA-2009: Send email when user request to join private beta

### PR DESCRIPTION
**Pull request by HarshVasudev**
- Backend lambda function to publish email message to a topic
- Addition of the publishToSNS function in the LambdaFacade
- Calling the publishToSNS from the express and front end /private-beta route on submit
- Saml templating of the SNS topic creation and its arn referencing

### CODE REVIEW

File: express/package.json [hunk 1]
```diff
@@ -18,36 +18,37 @@
   },
   "dependencies": {
     "@aws-sdk/client-cognito-identity-provider": "^3.80.0",
+    "@aws-sdk/client-sns": "^3.162.0",
```
Comment (CharlesIC) [inline express/package.json:21]: Can you please undo the changes to rearrange this file. The change should just be one line to add `"@aws-sdk/client-sns"`

Comment (HarshVasudev) [reply express/package.json:21]: Sure.

File: backend/dynamo-api/template.yaml [hunk 2]
```diff
@@ -9,6 +9,10 @@ Parameters:
   DynamoDbTableStackName:
     Type: String
     Description: The name of the stack exporting the table we want
+  PrivateBetaNotificationEmail:
+    Type: String
+    Description: Email address for private beta notification.
```
Comment (CharlesIC) [inline backend/dynamo-api/template.yaml:14]:
```suggestion
    Description: Email address for private beta request notifications
```

File: express/package.json (hunk 1 above)
Comment (CharlesIC) [inline express/package.json:21]: The express app doesn't need to have this package, or any other new packages, so the changes to this file should be reverted.

File: backend/dynamo-api/package.json [hunk 3]
```diff
@@ -11,14 +11,15 @@
...
+    "@aws-sdk/client-sns": "^3.162.0",
     "@aws-sdk/client-sts": "^3.160.0",
     "@aws-sdk/util-dynamodb": "^3.160.0",
     "axios": "^0.27.2"
   },
   "devDependencies": {
     "@types/aws-lambda": "^8.10.102",
-    "@types/node": "^18.7.14",
     "@types/jest": "^29.0.0",
+    "@types/node": "^18.7.14",
```
Comment (CharlesIC) [inline backend/dynamo-api/package.json:22]: Revert

**Review by CharlesIC — COMMENTED** [review]
This looks good, just a few minor changes are needed, and perhaps some tests.

File: express/package.json (hunk 1 above)
Comment (HarshVasudev) [reply express/package.json]: Done.

File: backend/dynamo-api/src/handlers/publish-to-sns.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import {APIGatewayProxyEvent, APIGatewayProxyResult} from 'aws-lambda';
+import { SNSClient, PublishCommand } from "@aws-sdk/client-sns";
+const snsClient = new SNSClient({ region: "eu-west-2" });
+
+export const publishToSNSHandler = async (event: APIGatewayProxyEvent, context: Context): Promise<APIGatewayProxyResult> => {
+    const payload = event?.body ? JSON.parse(event.body as string) : event;
+    console.debug(payload)
```
Comment (CharlesIC) [inline backend/dynamo-api/src/handlers/publish-to-sns.ts:7]:
```suggestion
```
Delete logging for now

File: backend/dynamo-api/src/handlers/publish-to-sns.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    const emailContent = 'Request to join GOV.UK Sign In private beta form submitted\n\n'+
+                         'Name: '+ payload.name + '\n' +
+                         'Email address: '+ payload.emailAddress + '\n' +
+                         'Service name: '+ payload.serviceName + '\n' +
+                         'Department: ' + payload.department + '\n'
```
Comment (CharlesIC) [inline backend/dynamo-api/src/handlers/publish-to-sns.ts:12]:
```suggestion
    const emailContent = "Request to join GOV.UK Sign In private beta form submitted\n\n" +
                         `Name: ${payload.name}\n` +
                         ...
```
Use string interpolation

File: backend/dynamo-api/template.yaml [hunk 6]
```diff
@@ -465,6 +469,42 @@ Resources:
...
+          NODE_OPTIONS: --enable-source-maps
+          SNS_TOPIC_ARN: !Ref PrivateBetaSnsTopic
+      Policies:
+        - SNSPublishMessagePolicy:
+            TopicName: !GetAtt PrivateBetaSnsTopic.TopicName
+      Events:
+        Api:
+          Type: Api
+          Properties:
+            Path: /publish-to-sns
```
Comment (CharlesIC) [inline backend/dynamo-api/template.yaml:497]:
```suggestion
            Path: /send-private-beta-request-notification
```

**Review by CharlesIC — DISMISSED** [review]

**Review by CharlesIC — APPROVED** [review]

**Review by CharlesIC — APPROVED** [review]
